subgraphs/notional-finance/src/mappings/notional.ts

```diff
@@ -0,0 +1,338 @@
+import { addToArrayAtIndex, removeFromArrayAtIndex } from "../common/arrays";
+
+export function handleLendBorrowTrade(event: LendBorrowTrade): void {
+  // TODO: why is nonce null? reading nonce value results in subgraph error
+  // log.error(" -- Before Nonce", []);
+  // let nonce = event.transaction.nonce;
+  // log.error(" -- After Nonce", [nonce.toString()]);
+
+  let currencyId = event.params.currencyId;
+  let maturity = event.params.maturity;
+  let marketId = currencyId.toString() + "-" + maturity.toString();
+
+  let market = getOrCreateMarket(event, marketId);
+  let account = getOrCreateAccount(event.params.account.toHexString(), event);
+  let token = getTokenFromCurrency(event, currencyId.toString());
+
+  // update market
+  market.inputTokenPriceUSD = token.lastPriceUSD!;
+  //
+  // TODO: market.rates cannot be fetched without settlement date
+  // let interestRate = getOrCreateInterestRate(market.id);
+  // let settlement = event.params.settlementDate;    // not available
+  // Get 'impliedRate' from 'getMarket(currencyId, maturity, settlementDate);
+  // interestRate.rate = impliedRate;
+  // market.rates = [interestRate];
+  //
+  // TODO: market exchange rate, talking to notional on how to find this
+  // market.exchangeRate
+  //
+  market.save();
+
+  // marketstatus entity
+  // TODO: we could make this protocol entity
+  // TODO: do this getOrCreateMarket?
+  let markets = getMarketsWithStatus(event);
+  if (markets.activeMarkets.indexOf(market.id) < 0) {
+    markets.activeMarkets = addToArrayAtIndex(
+      markets.activeMarkets,
+      market.id,
+      0
+    );
+    markets.save();
+  }
+
+  // getActiveMarkets
+  let currencyIds = [1, 2, 3, 4];
+  let notional = Notional.bind(Address.fromString(PROTOCOL_ID));
+  let activeMarkets: string[] = [];
+  for (let i = 0; i < currencyIds.length; i++) {
+    let call = notional.try_getActiveMarkets(currencyIds[i]);
+    if (call.reverted) {
+      log.error(
+        "[handleLendBorrowTrade] getActiveMarkets for currencyId {} reverted",
+        [currencyId.toString()]
+      );
+    } else {
+      for (let j = 0; j < call.value.length; j++) {
+        let currencyMarket =
+          currencyIds[i].toString() + "-" + call.value[j].maturity.toString();
+        activeMarkets = activeMarkets.concat([currencyMarket]);
+      }
+    }
+  }
+
+  for (let k = 0; k < markets.activeMarkets.length; k++) {
+    if (!activeMarkets.includes(markets.activeMarkets[k])) {
+      // event is irrelevant
+      let m = getOrCreateMarket(event, markets.activeMarkets[k]);
+
+      // status
+      m.isActive = false;
+      m.canBorrowFrom = false;
+
+      // balances
+      m.totalValueLockedUSD = BIGDECIMAL_ZERO;
+      m.totalDepositBalanceUSD = BIGDECIMAL_ZERO;
+      m.totalBorrowBalanceUSD = BIGDECIMAL_ZERO;
+
+      // positions
+      // TODO: should we update position counts manually to 0? do we need to manually close positions as markets mature?
```

Comment:
Interesting point. I'm assuming no event is triggered in any way when a position is closed, right?

Can you think of some way of updating these counters?

Reply:
Added a note in the readme. 

Summary, when a market matures:
- either the borrower rolls it over to the new market - in which case two individual trades happen 1) lending to marketing that is maturing and 2) borrowing from the new market. There is no special rollover function.
- or, a 3p settles the debt in the maturing market and so there isn't a new position in the new market.

Hence we should be fine with manually setting the positions to 0 here.

https://docs.notional.finance/notional-v2/risk-and-collateralization/settlement#third-party-settlement